### Apriori

#### importing the libraries

In [1]:
%pip install apyori

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
from apyori import apriori
import matplotlib.pyplot as plt
import numpy as np

#### Data Preprocessing

In [ ]:
dataset = pd.read_csv('../Datasets/Market_Basket_Optimisation.csv', header=None)
transactions = []
for i in range(0, 7501):
    transactions.append([str(dataset.values[i, j]) for j in range(0, 20)])

print(transactions)

#### Traing the Apriori model on the dataset

In [5]:
from apyori import apriori
rules = apriori(transactions=transactions, min_support=0.003, min_confidence=0.2, min_lift=3, min_length=2, max_length=2)

#### Visualizing the results

##### Displaying the first results coming directly from the output of the apriori function

In [ ]:
results = list(rules)
print(results)

##### Putting the results well organised into a Pandas DataFrame

In [7]:
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results]
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results), columns=['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])

##### Displaying the results non sorted

In [8]:
resultsinDataFrame

Left Hand Side Right Hand Side   Support  Confidence      Lift
0           light cream         chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
2                 pasta        escalope  0.005866    0.372881  4.700812
3         fromage blanc           honey  0.003333    0.245098  5.164271
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
6           light cream       olive oil  0.003200    0.205128  3.114710
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
8                 pasta          shrimp  0.005066    0.322034  4.506672

##### Displaying the sorted results

In [9]:
resultsinDataFrame.nlargest(n=10, columns='Lift')

Left Hand Side Right Hand Side   Support  Confidence      Lift
3         fromage blanc           honey  0.003333    0.245098  5.164271
0           light cream         chicken  0.004533    0.290598  4.843951
2                 pasta        escalope  0.005866    0.372881  4.700812
8                 pasta          shrimp  0.005066    0.322034  4.506672
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
6           light cream       olive oil  0.003200    0.205128  3.114710

#### Summary of Apriori Model

The **Apriori algorithm** is a popular method used in **Market Basket Analysis** to discover associations between items in large datasets. It helps answer questions like: *"If a customer buys item A, how likely are they to buy item B?"*

---

**Key Terms:**

- **Transaction:** A list of items purchased together (e.g., `['milk', 'bread', 'eggs']`).
- **Itemset:** A group of items (e.g., `{milk, bread}`).
- **Support:** The proportion of transactions that contain a particular itemset.  
    *Example:* If 3 out of 10 transactions contain `{milk, bread}`, support = 0.3.
- **Confidence:** The likelihood that item B is bought when item A is bought.  
    *Example:* If `{milk}` appears in 5 transactions and `{milk, bread}` in 3, confidence of `{milk} → {bread}` = 3/5 = 0.6.
- **Lift:** How much more likely item B is bought when item A is bought, compared to random chance.  
    *Lift > 1* means items are positively associated.

---

**How the Model Works:**

1. **Data Preparation:**  
     The dataset is transformed into a list of transactions, each representing items bought together.

2. **Applying Apriori:**  
     The algorithm scans the transactions to find frequent itemsets based on minimum support, confidence, and lift thresholds.

3. **Generating Rules:**  
     Association rules are created, showing relationships like `{milk} → {bread}`.

4. **Results Visualization:**  
     The strongest rules are displayed in a table, sorted by lift.

---

**Example Output Table:**

| Left Hand Side      | Right Hand Side | Support  | Confidence | Lift    |
|---------------------|-----------------|----------|------------|---------|
| light cream         | chicken         | 0.0045   | 0.29       | 4.84    |
| mushroom cream sauce| escalope        | 0.0057   | 0.30       | 3.79    |
| pasta               | escalope        | 0.0058   | 0.37       | 4.70    |

---